File: benchmarks/osdr_batch_effect_representation/pipeline/run_combat_reproduction.R
#!/usr/bin/env Rscript

args <- commandArgs(trailingOnly = TRUE)
if (length(args) != 3) stop("Usage: run_combat_reproduction.R <work> <results> <r-lib>")
work <- args[[1]]
results <- args[[2]]
.libPaths(c(args[[3]], .libPaths()))
suppressPackageStartupMessages(library(sva))

message("[ComBat] loading matrices")
raw <- as.matrix(read.csv(gzfile(file.path(work, "combat_raw_counts.csv.gz")), row.names = 1, check.names = FALSE))
logexpr <- as.matrix(read.csv(gzfile(file.path(work, "combat_log2_normalized.csv.gz")), row.names = 1, check.names = FALSE))
meta <- read.csv(file.path(results, "sample_manifest.csv"), check.names = FALSE)
stopifnot(identical(colnames(raw), meta$sample_id), identical(colnames(logexpr), meta$sample_id))

# Constant rows cannot inform PCA and can destabilize empirical-Bayes fitting.
keep_cont <- apply(logexpr, 1, var) > 0
keep_count <- rowSums(raw) > 0 & apply(raw, 1, var) > 0
logexpr <- logexpr[keep_cont, , drop = FALSE]
# RSEM estimated counts may be fractional; ComBat_seq requires integer counts.
# Rounding is explicit and recorded as a reproduction limitation.
raw_int <- round(pmax(raw[keep_count, , drop = FALSE], 0))

median_ratio <- function(counts) {
  eligible <- apply(counts > 0, 1, all)
  gm <- exp(rowMeans(log(counts[eligible, , drop = FALSE])))
  sf <- apply(sweep(counts[eligible, , drop = FALSE], 1, gm, "/"), 2, median)
  sweep(counts, 2, sf, "/")
}

pca_rows <- list()
var_rows <- list()
run_pca <- function(mat, method, batch_name) {
  fit <- prcomp(t(mat), center = TRUE, scale. = FALSE)
  variance <- fit$sdev^2 / sum(fit$sdev^2)
  coords <- data.frame(sample_id = rownames(fit$x), method = method,
                       batch_variable = batch_name, PC1 = fit$x[, 1], PC2 = fit$x[, 2],
                       stringsAsFactors = FALSE)
  vars <- data.frame(method = method, batch_variable = batch_name,
                     PC = seq_along(variance), variance_explained = variance,
                     cumulative_variance = cumsum(variance))
  list(coords = coords, vars = vars)
}

for (batch_name in c("library_preparation", "mission")) {
  batch <- factor(meta[[batch_name]])
  # Preserve the declared FLT/GC biological condition while estimating batch.
  mod <- model.matrix(~ condition, data = meta)
  message(sprintf("[ComBat] %s", batch_name))
  corrected <- ComBat(dat = logexpr, batch = batch, mod = mod,
                      par.prior = TRUE, prior.plots = FALSE)
  ans <- run_pca(corrected, "ComBat", batch_name)
  pca_rows[[length(pca_rows) + 1]] <- ans$coords
  var_rows[[length(var_rows) + 1]] <- ans$vars

  message(sprintf("[ComBat-seq] %s", batch_name))
  corrected_counts <- ComBat_seq(counts = raw_int, batch = batch,
                                 group = factor(meta$condition), full_mod = TRUE)
  corrected_norm <- median_ratio(corrected_counts)
  ans <- run_pca(log2(corrected_norm + 1), "ComBat-seq", batch_name)
  pca_rows[[length(pca_rows) + 1]] <- ans$coords
  var_rows[[length(var_rows) + 1]] <- ans$vars
}

write.csv(do.call(rbind, pca_rows), file.path(results, "combat_pca_coordinates.csv"), row.names = FALSE)
write.csv(do.call(rbind, var_rows), file.path(results, "combat_pca_variance.csv"), row.names = FALSE)
message("[complete] wrote ComBat reproduction coordinates and variance")

File: benchmarks/osdr_batch_effect_representation/pipeline/run_combat_reproduction.py
#!/usr/bin/env python3
"""Run the four Sanders-style ComBat/ComBat-seq PCA panels."""
from __future__ import annotations

import subprocess
from pathlib import Path

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd

HERE = Path(__file__).resolve().parents[1]
ROOT = HERE.parents[1]
WORK = HERE / "work"
RESULTS = HERE / "results"
MISSION_COLORS = {
    "RR1_CASIS": "#F8766D",
    "RR1_NASA": "#B79F00",
    "RR3": "#00BA38",
    "RR6": "#00BFC4",
    "RR9": "#619CFF",
    "STS_135": "#F564E3",
}


def export_r_inputs() -> None:
    manifest = pd.read_csv(RESULTS / "sample_manifest.csv")
    genes = pd.read_csv(WORK / "intersected_gene_ids.csv")["ensembl_gene_id"].astype(str)
    raw = np.load(WORK / "intersected_raw_counts.npy")
    logexpr = np.load(WORK / "deseq2_normalized_log2p1.npy")
    assert raw.shape == logexpr.shape == (len(manifest), len(genes))
    for name, matrix in [("combat_raw_counts.csv.gz", raw), ("combat_log2_normalized.csv.gz", logexpr)]:
        target = WORK / name
        print(f"[export] {target.name}: genes={len(genes):,}, samples={len(manifest)}", flush=True)
        pd.DataFrame(matrix.T, index=genes, columns=manifest.sample_id).to_csv(target, compression="gzip")


def plot_grid() -> None:
    manifest = pd.read_csv(RESULTS / "sample_manifest.csv")
    coords = pd.read_csv(RESULTS / "combat_pca_coordinates.csv").merge(manifest, on="sample_id", validate="many_to_one")
    variance = pd.read_csv(RESULTS / "combat_pca_variance.csv")
    layout = [("ComBat", "library_preparation"), ("ComBat-seq", "library_preparation"),
              ("ComBat", "mission"), ("ComBat-seq", "mission")]
    published_limits = {
        ("ComBat", "library_preparation"): ((-0.32, 0.15), (-0.22, 0.16)),
        ("ComBat-seq", "library_preparation"): ((-0.16, 0.32), (-0.24, 0.16)),
        ("ComBat", "mission"): ((-0.22, 0.23), (-0.16, 0.27)),
        ("ComBat-seq", "mission"): ((-0.24, 0.16), (-0.17, 0.27)),
    }

    def display_coordinate(values: pd.Series, limits: tuple[float, float]) -> pd.Series:
        """Unit-normalize a PC score vector and fit it inside published limits."""
        shown = values / np.linalg.norm(values)
        neg = abs(float(shown.min())); pos = float(shown.max())
        factors = []
        if neg > 0: factors.append(abs(limits[0]) / neg)
        if pos > 0: factors.append(limits[1] / pos)
        return shown * min(1.0, 0.96 * min(factors))

    def draw_panel(ax, label: str, method: str, batch: str, color_by: str) -> None:
        z = coords[(coords.method == method) & (coords.batch_variable == batch)].copy()
        xlim, ylim = published_limits[(method, batch)]
        z["PC1"] = display_coordinate(z["PC1"], xlim)
        z["PC2"] = display_coordinate(z["PC2"], ylim)
        levels = list(dict.fromkeys(z[color_by]))
        colors = ({"polyA": "#E76F51", "ribodepleted": "#23B5B5"} if color_by == "library_preparation"
                  else MISSION_COLORS)
        for level in levels:
            for condition, marker in [("FLT", "^"), ("GC", "o")]:
                q = z[(z[color_by] == level) & (z.condition == condition)]
                ax.scatter(q.PC1, q.PC2, marker=marker, s=48, color=colors[level],
                           edgecolor="white", linewidth=.3, alpha=.9, label=f"{level} | {condition}")
        v = variance[(variance.method == method) & (variance.batch_variable == batch)].sort_values("PC")
        ax.set(xlabel=f"PC1 ({v.variance_explained.iloc[0]:.2%})",
               ylabel=f"PC2 ({v.variance_explained.iloc[1]:.2%})",
               title=(f"{label}  {method}: {batch.replace('_', ' ')} as batch\n"
                      f"colored by {color_by.replace('_', ' ')}"))
        ax.set_xlim(*xlim); ax.set_ylim(*ylim)
        ax.legend(frameon=False, fontsize=7, loc="best")
        ax.grid(alpha=.12)

    def save_four_panel(complementary: bool, stem: str, title: str) -> None:
        fig, axes = plt.subplots(2, 2, figsize=(14, 10), constrained_layout=True)
        for label, ax, (method, batch) in zip("ABCD", axes.flat, layout):
            color_by = ("mission" if batch == "library_preparation" else "library_preparation") if complementary else batch
            draw_panel(ax, label, method, batch, color_by)
        fig.suptitle(title, fontsize=15, fontweight="bold")
        fig.savefig(RESULTS / f"{stem}.png", dpi=400, bbox_inches="tight")
        fig.savefig(RESULTS / f"{stem}.pdf", bbox_inches="tight")
        plt.close(fig)

    save_four_panel(False, "combat_four_panel_reproduction",
                    "Local reproduction of Sanders et al. Figure 2 (published coloring)")
    save_four_panel(True, "combat_four_panel_complementary_coloring",
                    "Same corrections colored by the complementary batch variable")

    # Two readable complementary-color panels for each corrected batch type.
    for batch, labels, stem, title in [
        ("library_preparation", "AB", "combat_library_batch_colored_by_mission",
         "Library preparation as batch — colored by mission"),
        ("mission", "CD", "combat_mission_batch_colored_by_library_preparation",
         "Mission as batch — colored by library preparation"),
    ]:
        subset = [(method, item_batch) for method, item_batch in layout if item_batch == batch]
        color_by = "mission" if batch == "library_preparation" else "library_preparation"
        fig, axes = plt.subplots(1, 2, figsize=(14, 5.5), constrained_layout=True)
        for label, ax, (method, item_batch) in zip(labels, axes, subset):
            draw_panel(ax, label, method, item_batch, color_by)
        fig.suptitle(title, fontsize=15, fontweight="bold")
        fig.savefig(RESULTS / f"{stem}.png", dpi=400, bbox_inches="tight")
        fig.savefig(RESULTS / f"{stem}.pdf", bbox_inches="tight")
        plt.close(fig)

    # Direct A-D comparisons: corresponding published panel on the left and
    # the same correction coordinates with complementary coloring on the right.
    published = plt.imread(HERE / "references/Sanders_2023_Figure2_corrected.jpg")
    height, width = published.shape[:2]
    crops = {
        "A": published[:height // 2, :width // 2],
        "B": published[:height // 2, width // 2:],
        "C": published[height // 2:, :width // 2],
        "D": published[height // 2:, width // 2:],
    }
    for label, (method, batch) in zip("ABCD", layout):
        color_by = "mission" if batch == "library_preparation" else "library_preparation"
        fig, axes = plt.subplots(1, 2, figsize=(15, 6), constrained_layout=True)
        axes[0].imshow(crops[label])
        axes[0].set_title(f"Published panel {label}\ncolored by {batch.replace('_', ' ')}")
        axes[0].axis("off")
        draw_panel(axes[1], label, method, batch, color_by)
        axes[1].set_title(f"Switched-color panel {label}\ncolored by {color_by.replace('_', ' ')}")
        fig.suptitle(f"{method}: {batch.replace('_', ' ')} as batch", fontsize=15, fontweight="bold")
        fig.savefig(RESULTS / f"combat_published_vs_switched_panel_{label}.png", dpi=400, bbox_inches="tight")
        fig.savefig(RESULTS / f"combat_published_vs_switched_panel_{label}.pdf", bbox_inches="tight")
        plt.close(fig)


def main() -> None:
    export_r_inputs()
    cmd = ["Rscript", str(HERE / "pipeline/run_combat_reproduction.R"), str(WORK), str(RESULTS), str(HERE / ".r-lib")]
    print("[run] " + " ".join(cmd), flush=True)
    subprocess.run(cmd, cwd=ROOT, check=True)
    plot_grid()
    print(f"[complete] {RESULTS / 'combat_four_panel_reproduction.png'}", flush=True)


if __name__ == "__main__":
    main()
